File: ios/TinyLLM/Engine/TinyGPT.swift
// On-device inference for the GPT trained by python/train.py.
//
// The math mirrors python/verify_export.py line by line, but processes one
// token at a time with a per-layer KV cache, so each generated token costs a
// single forward pass over the new position only. When the context window
// fills up, the cache is rebuilt from the most recent half of the window
// (learned absolute positions make cached entries position-dependent, so a
// plain sliding window would be wrong).

import Foundation

// @unchecked Sendable: the generation pipeline touches a TinyGPT from one
// task at a time (Generator serializes access); the annotation keeps the
// engine usable from a background Task under Swift 6 strict concurrency.
final class TinyGPT: @unchecked Sendable {
    let config: ModelConfig
    private let w: [String: [Float]]
    private let headDim: Int

    // kCache[layer] / vCache[layer] hold one n_embd-length k/v vector per
    // cached position, concatenated; head h of position t lives at
    // [t * n_embd + h * headDim ..< t * n_embd + (h+1) * headDim].
    private var kCache: [[Float]]
    private var vCache: [[Float]]
    private(set) var cachedTokens: [Int] = []

    init(checkpoint: Checkpoint) {
        self.config = checkpoint.config
        self.w = checkpoint.tensors
        self.headDim = config.n_embd / config.n_head
        self.kCache = Array(repeating: [], count: config.n_layer)
        self.vCache = Array(repeating: [], count: config.n_layer)
    }

    func reset() {
        for i in 0..<config.n_layer {
            kCache[i].removeAll(keepingCapacity: true)
            vCache[i].removeAll(keepingCapacity: true)
        }
        cachedTokens.removeAll(keepingCapacity: true)
    }

    /// Feed one token; returns the logits for the next token.
    func step(_ token: Int) -> [Float] {
        if cachedTokens.count == config.block_size {
            // context full: re-prefill from the most recent half of the window
            let tail = Array(cachedTokens.suffix(config.block_size / 2))
            reset()
            for t in tail { _ = forwardOne(t) }
        }
        return forwardOne(token)
    }

    private func forwardOne(_ token: Int) -> [Float] {
        let E = config.n_embd
        let pos = cachedTokens.count

        var x = [Float](repeating: 0, count: E)
        let tokEmb = w["tok_emb.weight"]!
        let posEmb = w["pos_emb.weight"]!
        for i in 0..<E {
            x[i] = tokEmb[token * E + i] + posEmb[pos * E + i]
        }

        for layer in 0..<config.n_layer {
            let p = "blocks.\(layer)."

            // -- causal self-attention over the cache --
            var h = x
            MathOps.layerNorm(&h, gain: w[p + "ln1.weight"]!, bias: w[p + "ln1.bias"]!)
            let qkv = MathOps.matVec(w[p + "attn.qkv.weight"]!, rows: 3 * E, cols: E,
                                     h, bias: w[p + "attn.qkv.bias"]!)
            let q = Array(qkv[0..<E])
            kCache[layer].append(contentsOf: qkv[E..<(2 * E)])
            vCache[layer].append(contentsOf: qkv[(2 * E)..<(3 * E)])
            let T = pos + 1

            var attnOut = [Float](repeating: 0, count: E)
            let scale = 1.0 / sqrt(Float(headDim))
            q.withUnsafeBufferPointer { qp in
                kCache[layer].withUnsafeBufferPointer { kp in
                    vCache[layer].withUnsafeBufferPointer { vp in
                        var scores = [Float](repeating: 0, count: T)
                        for head in 0..<config.n_head {
                            let hOff = head * headDim
                            for t in 0..<T {
                                scores[t] = MathOps.dot(qp.baseAddress! + hOff,
                                                        kp.baseAddress! + t * E + hOff,
                                                        headDim) * scale
                            }
                            MathOps.softmax(&scores)
                            for t in 0..<T {
                                let prob = scores[t]
                                let vBase = t * E + hOff
                                for d in 0..<headDim {
                                    attnOut[hOff + d] += prob * vp[vBase + d]
                                }
                            }
                        }
                    }
                }
            }
            let attnProj = MathOps.matVec(w[p + "attn.proj.weight"]!, rows: E, cols: E,
                                          attnOut, bias: w[p + "attn.proj.bias"]!)
            for i in 0..<E { x[i] += attnProj[i] }

            // -- MLP --
            h = x
            MathOps.layerNorm(&h, gain: w[p + "ln2.weight"]!, bias: w[p + "ln2.bias"]!)
            var mid = MathOps.matVec(w[p + "mlp.fc.weight"]!, rows: 4 * E, cols: E,
                                     h, bias: w[p + "mlp.fc.bias"]!)
            MathOps.gelu(&mid)
            let mlpOut = MathOps.matVec(w[p + "mlp.proj.weight"]!, rows: E, cols: 4 * E,
                                        mid, bias: w[p + "mlp.proj.bias"]!)
            for i in 0..<E { x[i] += mlpOut[i] }
        }

        MathOps.layerNorm(&x, gain: w["ln_f.weight"]!, bias: w["ln_f.bias"]!)
        // weight-tied LM head: logits = tok_emb · x
        let logits = MathOps.matVec(tokEmb, rows: config.vocab_size, cols: E, x)

        cachedTokens.append(token)
        return logits
    }
}

// MARK: - Sampling

struct Sampler {
    var temperature: Float = 0.8
    var topK: Int = 40

    func sample(_ logits: [Float]) -> Int {
        var l = logits
        let temp = max(temperature, 1e-6)
        for i in 0..<l.count { l[i] /= temp }

        if topK > 0 && topK < l.count {
            let threshold = l.sorted(by: >)[topK - 1]
            for i in 0..<l.count where l[i] < threshold { l[i] = -.greatestFiniteMagnitude }
        }

        MathOps.softmax(&l)
        var r = Float.random(in: 0..<1)
        for (i, p) in l.enumerated() {
            r -= p
            if r < 0 { return i }
        }
        return l.count - 1
    }
}

// MARK: - Streaming generation

final class Generator: @unchecked Sendable {
    private let model: TinyGPT
    private let tokenizer: CharTokenizer

    init(model: TinyGPT, tokenizer: CharTokenizer) {
        self.model = model
        self.tokenizer = tokenizer
    }

    /// Streams generated text, one decoded token at a time.
    func generate(prompt: String, maxNewTokens: Int, sampler: Sampler) -> AsyncStream<String> {
        AsyncStream { continuation in
            let task = Task.detached(priority: .userInitiated) { [model, tokenizer] in
                model.reset()
                var ids = tokenizer.encode(prompt)
                if ids.isEmpty { ids = [0] }

                // prefill: feed the prompt, keep only the final logits
                var logits: [Float] = []
                for id in ids {
                    if Task.isCancelled { continuation.finish(); return }
                    logits = model.step(id)
                }

                for _ in 0..<maxNewTokens {
                    if Task.isCancelled { break }
                    let next = sampler.sample(logits)
                    continuation.yield(tokenizer.decode([next]))
                    logits = model.step(next)
                }
                continuation.finish()
            }
            continuation.onTermination = { _ in task.cancel() }
        }
    }
}

File: web/tinygpt.js
// TinyGPT inference in pure JavaScript — the third sibling of
// python/verify_export.py (NumPy) and ios/TinyLLM/Engine/TinyGPT.swift.
// Loads the same exported model files and produces the same logits.
// KV-cached single-token stepping; no dependencies.

// Abramowitz & Stegun 7.1.26 erf approximation (max error 1.5e-7),
// close enough to the exact-erf GELU used in training
function erf(x) {
  const sign = x < 0 ? -1 : 1;
  x = Math.abs(x);
  const t = 1 / (1 + 0.3275911 * x);
  const y = 1 - (((((1.061405429 * t - 1.453152027) * t) + 1.421413741) * t
    - 0.284496736) * t + 0.254829592) * t * Math.exp(-x * x);
  return sign * y;
}

const SQRT2 = Math.sqrt(2);

function gelu(x, n) {
  for (let i = 0; i < n; i++) x[i] = 0.5 * x[i] * (1 + erf(x[i] / SQRT2));
}

function layerNorm(x, n, gain, bias) {
  let mean = 0;
  for (let i = 0; i < n; i++) mean += x[i];
  mean /= n;
  let variance = 0;
  for (let i = 0; i < n; i++) { const d = x[i] - mean; variance += d * d; }
  variance /= n;
  const inv = 1 / Math.sqrt(variance + 1e-5);
  for (let i = 0; i < n; i++) x[i] = (x[i] - mean) * inv * gain[i] + bias[i];
}

function softmaxInPlace(x, n) {
  let max = -Infinity;
  for (let i = 0; i < n; i++) if (x[i] > max) max = x[i];
  let sum = 0;
  for (let i = 0; i < n; i++) { x[i] = Math.exp(x[i] - max); sum += x[i]; }
  for (let i = 0; i < n; i++) x[i] /= sum;
}

// out[r] = sum_c w[r*cols+c] * x[c] (+ bias[r]);  w is [rows x cols] row-major
function matVec(w, rows, cols, x, bias, out) {
  for (let r = 0; r < rows; r++) {
    let acc = 0;
    const base = r * cols;
    for (let c = 0; c < cols; c++) acc += w[base + c] * x[c];
    out[r] = bias ? acc + bias[r] : acc;
  }
}

export class CharTokenizer {
  constructor(chars) {
    this.chars = chars;
    this.stoi = new Map(chars.map((c, i) => [c, i]));
  }
  encode(text) {
    const ids = [];
    for (const ch of text) {
      const id = this.stoi.get(ch);
      if (id !== undefined) ids.push(id);
    }
    return ids;
  }
  decode(ids) { return ids.map(i => this.chars[i]).join(""); }
}

export class TinyGPT {
  constructor(config, weights) {
    this.cfg = config;
    this.w = weights; // {name: Float32Array}
    this.headDim = config.n_embd / config.n_head;
    const E = config.n_embd;
    // scratch buffers
    this.x = new Float32Array(E);
    this.h = new Float32Array(E);
    this.qkv = new Float32Array(3 * E);
    this.attnOut = new Float32Array(E);
    this.proj = new Float32Array(E);
    this.mid = new Float32Array(4 * E);
    this.scores = new Float32Array(config.block_size);
    this.logits = new Float32Array(config.vocab_size);
    // per-layer KV caches: [block_size x E], head h of position t at t*E + h*hd
    this.kCache = [];
    this.vCache = [];
    for (let l = 0; l < config.n_layer; l++) {
      this.kCache.push(new Float32Array(config.block_size * E));
      this.vCache.push(new Float32Array(config.block_size * E));
    }
    this.cached = [];
  }

  reset() { this.cached.length = 0; }

  // feed one token, return logits for the next (Float32Array, reused!)
  step(token) {
    if (this.cached.length === this.cfg.block_size) {
      // context full: re-prefill from the most recent half of the window
      const tail = this.cached.slice(-this.cfg.block_size / 2);
      this.reset();
      for (const t of tail) this._forwardOne(t);
    }
    return this._forwardOne(token);
  }

  _forwardOne(token) {
    const { cfg, w, headDim } = this;
    const E = cfg.n_embd;
    const pos = this.cached.length;
    const { x, h, qkv, attnOut, proj, mid, scores } = this;

    const tokEmb = w["tok_emb.weight"];
    const posEmb = w["pos_emb.weight"];
    for (let i = 0; i < E; i++) x[i] = tokEmb[token * E + i] + posEmb[pos * E + i];

    for (let l = 0; l < cfg.n_layer; l++) {
      const p = `blocks.${l}.`;

      // causal self-attention over the cache
      h.set(x);
      layerNorm(h, E, w[p + "ln1.weight"], w[p + "ln1.bias"]);
      matVec(w[p + "attn.qkv.weight"], 3 * E, E, h, w[p + "attn.qkv.bias"], qkv);
      const kC = this.kCache[l], vC = this.vCache[l];
      kC.set(qkv.subarray(E, 2 * E), pos * E);
      vC.set(qkv.subarray(2 * E, 3 * E), pos * E);
      const T = pos + 1;
      const scale = 1 / Math.sqrt(headDim);

      attnOut.fill(0);
      for (let head = 0; head < cfg.n_head; head++) {
        const hOff = head * headDim;
        for (let t = 0; t < T; t++) {
          let acc = 0;
          const kBase = t * E + hOff;
          for (let d = 0; d < headDim; d++) acc += qkv[hOff + d] * kC[kBase + d];
          scores[t] = acc * scale;
        }
        softmaxInPlace(scores, T);
        for (let t = 0; t < T; t++) {
          const prob = scores[t];
          const vBase = t * E + hOff;
          for (let d = 0; d < headDim; d++) attnOut[hOff + d] += prob * vC[vBase + d];
        }
      }
      matVec(w[p + "attn.proj.weight"], E, E, attnOut, w[p + "attn.proj.bias"], proj);
      for (let i = 0; i < E; i++) x[i] += proj[i];

      // MLP
      h.set(x);
      layerNorm(h, E, w[p + "ln2.weight"], w[p + "ln2.bias"]);
      matVec(w[p + "mlp.fc.weight"], 4 * E, E, h, w[p + "mlp.fc.bias"], mid);
      gelu(mid, 4 * E);
      matVec(w[p + "mlp.proj.weight"], E, 4 * E, mid, w[p + "mlp.proj.bias"], proj);
      for (let i = 0; i < E; i++) x[i] += proj[i];
    }

    layerNorm(x, E, w["ln_f.weight"], w["ln_f.bias"]);
    matVec(tokEmb, cfg.vocab_size, E, x, null, this.logits); // weight-tied head

    this.cached.push(token);
    return this.logits;
  }
}

export function sample(logits, temperature, topK, rng = Math.random) {
  const n = logits.length;
  const l = Float32Array.from(logits);
  if (temperature <= 0.05) {
    let best = 0;
    for (let i = 1; i < n; i++) if (l[i] > l[best]) best = i;
    return best;
  }
  for (let i = 0; i < n; i++) l[i] /= temperature;
  if (topK > 0 && topK < n) {
    const sorted = Float32Array.from(l).sort().reverse();
    const threshold = sorted[topK - 1];
    for (let i = 0; i < n; i++) if (l[i] < threshold) l[i] = -Infinity;
  }
  softmaxInPlace(l, n);
  let r = rng();
  for (let i = 0; i < n; i++) {
    r -= l[i];
    if (r < 0) return i;
  }
  return n - 1;
}

// parse weights.bin + manifest into named Float32Array views
export function sliceWeights(manifest, buffer) {
  const all = new Float32Array(buffer);
  const weights = {};
  for (const e of manifest) {
    weights[e.name] = all.subarray(e.offset, e.offset + e.count);
  }
  return weights;
}
